tensorflow/contrib/opt/python/training/elastic_average_optimizer.py

```diff
@@ -83,24 +103,50 @@ def __call__(self, getter, name, trainable, collections, *args, **kwargs):
             collections=[ops.GraphKeys.LOCAL_VARIABLES],
             *args,
             **kwargs)
-      global_center_variable = variable_scope.variable(
+      if kwargs['reuse'] == True:
+        return local_var
+      global_center_variable = getter(
           name='%s/%s' % (GLOBAL_VARIABLE_NAME, name),
-          initial_value=local_var.initialized_value(),
           trainable=False,
-          collections=[ops.GraphKeys.GLOBAL_VARIABLES])
+          collections=[ops.GraphKeys.GLOBAL_VARIABLES],
+          *args,
+          **kwargs)
 
       with ops.device(self._worker_device):
-        local_center_variable = variable_scope.variable(
+        local_center_variable = getter(
             name='%s/%s' % (LOCAL_VARIABLE_NAME, name),
-            initial_value=local_var.initialized_value(),
             trainable=False,
-            collections=[ops.GraphKeys.LOCAL_VARIABLES])
-
-      self._local_map[local_var] = local_center_variable
-      self._global_map[local_var] = global_center_variable
+            collections=[ops.GraphKeys.LOCAL_VARIABLES],
+            *args,
+            **kwargs)
+      if kwargs['partitioner'] is None:
+        self._local_map[local_var] = local_center_variable
+        self._global_map[local_var] = global_center_variable
+      else:
+        v_list = list(local_var)
+        for i in range(len(v_list)):
+          self._local_map[v_list[i]] \
+              = list(local_center_variable)[i]
+          self._global_map[v_list[i]] \
+              = list(global_center_variable)[i]
       return local_var
     else:
-      return getter(name, trainable, collections, *args, **kwargs)
+      # 1. default to LOCAL_VARIABLES (instead of GLOBAL_VARIABLES)
```

Comment:
This policy for variable placement seems really complicated and error-prone. Why is it being done?

Reply:
Assuming user code was like this in single GPU version:
```
with tf.variable_scope('model'):
    model = UserModel(...)
```
When using ElasticAverageOptimizer, code will be changed to like following:
```
with tf.device(tf.train.replica_device_setter(...)), \
        tf.variable_scope('model', custom_getter=custom_getter):
    model = UserModel(...)
```
Workers train their own local model and communicate with PS to update the 'global_center_variable' model. On single GPU version, variables defaults to GLOBAL_VARIABLES are fine, however for distributed version, they are logically belonging to local workers, so I move them to LOCAL_VARIABLES.
global_step is, of course, an exception, as it is shared across workers.
In case user does have cases similar to global_step, they can put the variables into the GLOBAL_SHARE_VARS collection to make it 'global'.

Reply:
Right. But setting these variables to LOCAL_VARIABLES will not affect their placement by a replica_device_setter, I think, so I don't understand what is being done here.

Reply:
Accept. Remove this part.
Verified on 4 different models, works good.

Reply:
I don't see it removed yet.

Reply:
sorry, just pushed the change.